client/state/data-layer/wpcom/login-2fa/index.js

```diff
@@ -20,77 +20,83 @@ import {
-			auth_type: authType,
-			two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
-			remember_me: getRememberMe( store.getState() ),
-			two_step_push_token: getTwoFactorPushToken( store.getState() ),
-			client_id: config( 'wpcom_signup_id' ),
-			client_secret: config( 'wpcom_signup_key' ),
-		} ).then( () => {
-			store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
-		} ).catch( error => {
-			store.dispatch( {
-				type: TWO_FACTOR_AUTHENTICATION_UPDATE_NONCE,
-				nonceType: 'push',
-				twoStepNonce: error.response.body.data.two_step_nonce
-			} );
-			return Promise.reject( error );
-		} );
+	store.dispatch(
+		http( {
+			url: 'https://wordpress.com/wp-login.php?action=two-step-authentication-endpoint',
+			method: 'POST',
+			headers: [
+				[ 'Content-Type', 'application/x-www-form-urlencoded' ],
+			],
+			withCredentials: true,
+			body: {
+				user_id: getTwoFactorUserId( store.getState() ),
+				auth_type: authType,
+				two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
+				remember_me: getRememberMe( store.getState() ),
+				two_step_push_token: getTwoFactorPushToken( store.getState() ),
+				client_id: config( 'wpcom_signup_id' ),
+				client_secret: config( 'wpcom_signup_key' ),
+			},
+		},
+		action
+	) );
 };
 
+const receivedTwoFactorPushNotificationApproved = ( { dispatch } ) =>
+	dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
+
 /***
- * Polling the login API for the status of the push notification.
- * The polling will stop on success or when store's polling progress
- * state changes to `false`
+ * Receive error from the two factor push notification status http request
  *
- * @param {Object}   store  Global redux store
+ * @param {Object}	store  Global redux store
+ * @param {Object}	action dispathced action
+ * @param {Object}	error the error object
  */
-const doAppPushPolling = store => {
-	let retryCount = 0;
-	const retry = () => {
-		// if polling was stopped or not in progress - stop
-		if ( ! getTwoFactorPushPollInProgress( store.getState() ) ) {
-			return;
-		}
+const receivedTwoFactorPushNotificationError = ( store, action, next, error ) => {
+	const twoStepNonce = get( error, 'response.body.data.two_step_nonce' );
 
-		retryCount++;
-		setTimeout(
-			() => doAppPushRequest( store ).catch( retry ),
-			( POLL_APP_PUSH_INTERVAL_SECONDS + Math.floor( retryCount / 10 ) ) * 1000 // backoff lineary
-		);
-	};
+	if ( ! twoStepNonce ) {
+		store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_STOP } ); //
+		throw new Error( "Two step nonce wasn't present on the response from polling endpoint, unable to continue" );
+	}
+
+	store.dispatch( {
+		type: TWO_FACTOR_AUTHENTICATION_UPDATE_NONCE,
+		nonceType: 'push',
+		twoStepNonce,
+	} );
 
 	if ( getTwoFactorPushPollInProgress( store.getState() ) ) {
-		doAppPushRequest( store ).catch( retry );
+		setTimeout(
+			// eslint-disable-next-line no-use-before-define
```

Comment:
Actually, after reviewing here I see another difference is that the UI _is_ coupled in here via the polling action types, which is what originally made me peek in on this. My suggestion related more to the idea of eliminating the `_POLL_START` and `_POLL_COMPLETED` actions as I felt those are superfluous.

If we want to cancel in progress than that's another question but I think it would be fair to have an explicit action type because that's an actual actionable item a customer might take - "I want to click this button to stop waiting for the response" and in that case we could still track it here so that when the response comes back we know not to reschedule it.